DRA: device taints and tolerations (#49822)
kubernetes/website: content/en/docs/concepts/scheduling-eviction/dynamic-resource-allocation.md

using a tainted device and to all Pods referencing that ResourceClaim.
The "NoSchedule" effect prevents scheduling those Pods.
Tainted devices are ignored when trying to allocate a ResourceClaim
because using them would prevent scheduling of Pods.

The "NoExecute" effect implies "NoSchedule" and in addition causes eviction
of all Pods which have been scheduled already. This eviction is implemented
in the device taint eviction controller in kube-controller-manager by
deleting affected Pods.

ResourceClaims can tolerate taints. If a taint is tolerated, its effect does
not apply. An empty toleration matches all taints. A toleration can be limited to
certain effects and/or match certain key/value pairs. A toleration can check
that a certain key exists, regardless which value it has, or it can check
for specific values of a key.
For more information on this matching see the
[node taint concepts](/docs/concepts/scheduling-eviction/taint-and-toleration#concepts).

Eviction can be delayed by tolerating a taint for a certain duration.
That delay starts at the time when a taint gets added to a device, which is recorded in a field
of the taint.

> Taints apply as described above also to ResourceClaims allocating "all" devices on a node.

Author: We have generated documentation for the API. I prefer to not repeat that manually. It's a bit hard to link to right now (new API hasn't been generated yet), which is why I haven't linked to it yet.

> Taints apply as described above also to ResourceClaims allocating "all" devices on a node.

Reviewer: Thanks, I see. Would you mind resoving this because it seems I can't.

All devices must be untainted or all of their taints must be tolerated.

> Allocating a device with admin access (described [below](#admin-access))

Reviewer 2: ```suggestion
Allocating a device with admin access (described [above](#admin-access))
```

is not exempt either. An admin using that mode must explicitly tolerate all taints
to access tainted devices.

Taints can be added to devices in two different ways:

### Taints set by the driver

A DRA driver can add taints to the device information that it publishes in ResourceSlices.
Consult the documentation of a DRA driver to learn whether the driver uses taints and what
their keys and values are.

### Taints set by an admin

An admin or a control plane component can taint devices without having to tell
the DRA driver to include taints in its ResourceSlices. They do that by

> creating ResourceSliceRules. Each ResourceSliceRule adds one taint to devices which

Reviewer 2: ```suggestion
creating DeviceTaintRules. Each DeviceTaintRule adds one taint to devices which
```

[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
is enabled in the kube-apiserver.

### Enabling device taints and tolerations

[Device taints and tolerations](#device-taints-and-tolerations) is an *alpha feature* and only enabled when the
`DRADeviceTaints` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)

> is enabled in the kube-apiserver and kube-scheduler. To use ResourceSliceRules, the

Reviewer: Yes, I think so (ref: https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/5055-dra-device-taints-and-tolerations/kep.yaml#L29-L36).

 ```suggestion
is enabled in the kube-apiserver, kube-scheduler and kube-controller-manager. To use DeviceTaintRules, the
```

> is enabled in the kube-apiserver and kube-scheduler. To use ResourceSliceRules, the

Author: Fixed, also via search/replace. I was pretty consistent with `ResourceSliceRule`...
